### PR TITLE
Add `Subarray`.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**

### CODE REVIEW

**Comment by teo-tsirpanis** [comment]
For now the Core does not support strides. Once it does, we will introduce `AddRange` overloads that take the stride as well, and `GetRangeAndStride` methods that return all three components.

File: sources/TileDB.CSharp/Subarray.cs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        public (T Start, T End) GetRange<T>(string dimensionName, uint rangeIndex) where T : struct
+        {
+            _ = EnumUtil.TypeToDataType(typeof(T));
+            using var ctxHandle = _ctx.Handle.Acquire();
+            using var handle = _handle.Acquire();
+            using var ms_dimensionName = new MarshaledString(dimensionName);
+            void* startPtr, endPtr, stridePtr_unused;
+            _ctx.handle_error(Methods.tiledb_subarray_get_range_from_name(ctxHandle, handle, ms_dimensionName, rangeIndex, &startPtr, &endPtr, &stridePtr_unused));
+            return (Unsafe.ReadUnaligned<T>(startPtr), Unsafe.ReadUnaligned<T>(endPtr));
+        }
```
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/Subarray.cs:310]: The equivalent `Query` methods did not validate types and neither do these; we could make them to now or in a future PR.

**Comment by teo-tsirpanis** [comment]
@shaunrd0 this is ready for review.

**Review by shaunrd0 — APPROVED** [review]
Looks great! A few minor comments but nothing blocking. :+1:

File: sources/TileDB.CSharp/Query.cs [hunk 2]
```diff
@@ -212,10 +213,54 @@ public Config Config()
...
+                }
+            }
+
+            return new Subarray(_ctx, _array, handle);
+        }
+
+        /// <summary>
+        /// Deprecated, use <see cref="Subarray.SetSubarray{T}(T[])"/> instead and assign the subarray with <see cref="SetSubarray(Subarray)"/>.
+        /// </summary>
+        [Obsolete(Obsoletions.QuerySubarrayMessage + " Use Subarray.SetSubarray instead on the query's assigned Subarray instead.", DiagnosticId = Obsoletions.QuerySubarrayDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
```
[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply sources/TileDB.CSharp/Query.cs]: Thanks for spotting it. Fixed.

File: tests/TileDB.CSharp.Test/AttributeTest.cs [hunk 3]
```diff
@@ -241,7 +241,9 @@ private void readVarLengthAttributeArray()
             var query_read = new Query(context, array_read);
 
             // Slice only rows 1,2 and cols 2,3,4
-            query_read.SetSubarray(new int[] { 1, 4, 1, 4});
+            var subarray = new Subarray(array_read);
```
[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply tests/TileDB.CSharp.Test/AttributeTest.cs]: Fixed.

File: tests/TileDB.CSharp.Test/QueryConditionTest.cs [hunk 4]
```diff
@@ -73,7 +73,8 @@ public void TestSimpleQueryCondition()
 
             query_read.SetLayout(LayoutType.RowMajor);
 
-            int[] subarray = new int[4] { 1, 4, 1, 4 };
+            var subarray = new Subarray(array_read);
```
[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply tests/TileDB.CSharp.Test/QueryConditionTest.cs]: Fixed.

**Comment by teo-tsirpanis** [comment]
I also used the new APIs in the delete tests (just mechanical changes).

CI is green, merging…

**Pull request merged**